skipchain/skipchain.go

```diff
@@ -863,105 +827,91 @@ func (s *Service) bftVerifyNewBlock(msg []byte, data []byte) (out bool) {
-		return errors.New("Can't have an index < 0")
-	}
-	if len(sb.BackLinkIDs) <= 0 {
-		return errors.New("Need at least one backlinkID")
-	}
-	if sb.Height < 1 {
-		return errors.New("Minimum height is 1")
+// forwardSignature receives a signature request of a newly accepted block.
+// It only needs the 2nd-newest block and the forward-link.
+func (s *Service) forwardSignature(fs *ForwardSignature) error {
+	if fs.TargetHeight >= len(fs.Newest.BackLinkIDs) {
+		return errors.New("This backlink-height doesn't exist")
 	}
-	if sb.Height > sb.MaximumHeight {
-		return errors.New("Height must be <= maximumHeight")
+	target := s.db.GetByID(fs.Newest.BackLinkIDs[fs.TargetHeight])
+	if target == nil {
+		return errors.New("Didn't find target-block")
 	}
-	if sb.Roster == nil {
-		return errors.New("Need a roster")
+	if !fs.Previous.Equal(target.Hash) {
+		return errors.New("TargetHeight backlink doesn't correspond to previous")
 	}
-	return nil
-}
-
-// addForwardLink verifies if the new block is valid. If it is not valid, it
-// returns with an error.
-// If it finds a valid block, a forward-link will be added and a BFT-signature
-// requested.
-func (s *Service) addForwardLink(src, dst *SkipBlock) error {
-	if src.GetForwardLen() > 0 {
-		return errors.New("already have forward-link at this height")
-	}
-
-	// create the message we want to sign for this round
-	roster := src.Roster
-	log.Lvlf3("%s is adding forward-link to %s: %d->%d", s.ServerIdentity(),
-		roster.List, src.Index, dst.Index)
-	data, err := network.Marshal(dst)
+	data, err := network.Marshal(fs)
 	if err != nil {
-		return fmt.Errorf("Couldn't marshal block: %s", err.Error())
+		return err
 	}
-	msg := []byte(dst.Hash)
-	sig, err := s.startBFT(bftNewBlock, roster, msg, append(src.Hash, data...))
+	fl := NewForwardLink(target, fs.Newest)
+	sig, err := s.startBFT(bftFollowBlock, target.Roster, fl.Hash(), data)
 	if err != nil {
-		log.Lvl3("startBFT failed with", err)
-		return err
+		return errors.New("Couldn't get signature: " + err.Error())
 	}
+	log.Lvl1("Adding forward-link to", target.Index)
 
-	fwd := &BlockLink{*sig}
-
-	// sanity check
-	if !dst.Hash.Equal(fwd.Hash()) {
-		panic("invalid message in signature")
+	fl.Signature = *sig
+	if !target.Roster.ID.Equal(fs.Newest.Roster.ID) {
+		fl.NewRoster = fs.Newest.Roster
 	}
+	target.AddForward(fl)
+	s.startPropagation([]*SkipBlock{target})
+	return nil
+}
 
-	fwl := s.db.GetByID(src.Hash).ForwardLink
-	log.Lvlf3("%s adds forward-link to %s: %d->%d - fwlinks:%v", s.ServerIdentity(),
-		roster.List, src.Index, dst.Index, fwl)
-	if len(fwl) > 0 {
-		return errors.New("forward-link got signed during our signing")
-	}
-	src.ForwardLink = []*BlockLink{fwd}
-	if err = src.VerifyForwardSignatures(); err != nil {
-		return errors.New("Wrong BFT-signature: " + err.Error())
+// verifyFollowBlock makes sure that a signature-request for a forward-link
+// is valid.
+func (s *Service) bftVerifyFollowBlock(msg []byte, data []byte) bool {
```

Comment:
I didn't see where the `From` part of the forward link is verified in this function. Is that somewhere else or there is no need to check it explicitly (as it's already a part of the hash and it's signed).

Reply:
It's done towards the end where a `ForwardLink` is created and then his Hash is compared to the final hash:

```
		fl := NewForwardLink(target, fs.Newest)
		if bytes.Compare(fl.Hash(), msg) != 0 {
```

So if the forward-link would be wrong, the hash would be changed.